{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aln2cov(samfile, ref_file):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"rb\")\n",
    "    refseq = SeqIO.read(ref_file, \"fasta\").seq\n",
    "    coverage = {}\n",
    "    \n",
    "    for read in samfile.fetch():\n",
    "        start_read, end_read = read.reference_start, read.reference_end\n",
    "        if start_read and end_read: # Check mapped\n",
    "            for pos in range(start_read, end_read):\n",
    "                if pos in coverage:\n",
    "                    coverage[pos] += 1\n",
    "                else:\n",
    "                    coverage[pos] = 1\n",
    "    samfile.close()\n",
    "    return coverage\n",
    "\n",
    "def mean_coverage(dict_coverage):\n",
    "    return np.array(list(dict_coverage.values())).mean()\n",
    "\n",
    "def count_coverage(dict_coverage, refrence_sequance):\n",
    "    return round(len(dict_coverage.keys()) / len(refrence_sequance), 15)\n",
    "\n",
    "def plot_coverage(dict_coverage):\n",
    "    x, y = list(dict_coverage.keys()), list(dict_coverage.values())\n",
    "    plt.figure(figsize=(15, 10))\n",
    "    plt.xlabel('Позиция нуклеотида в геноме')\n",
    "    plt.ylabel('Количество ридов, которые покрыли нуклеотид')\n",
    "    plt.axhline(y=mean_coverage(dict_coverage), label='Среднее покрытие', color='r', linestyle='--')\n",
    "    plt.legend()\n",
    "    sns.scatterplot(x, y)\n",
    "\n",
    "def count_indels(samfile):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"r\") \n",
    "    insertions, deletions, counter = 0, 0, 0\n",
    "    \n",
    "    for x in samfile.fetch():\n",
    "        if x.cigartuples:\n",
    "            for type_indel, lenght in x.cigartuples:\n",
    "                if type_indel == 0:\n",
    "                    counter += lenght\n",
    "                if type_indel == 1:\n",
    "                    insertions += lenght\n",
    "                if type_indel == 2:\n",
    "                    deletions += lenght\n",
    "    return (insertions / counter, deletions / counter)\n",
    "    \n",
    "def plot_indel(samfile):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"r\") \n",
    "    lenght_insertion = []\n",
    "    lenght_deletion = []\n",
    "    lenght_indel = []\n",
    "    for x in samfile.fetch():\n",
    "        if x.cigartuples:\n",
    "            for type_indel, lenght in x.cigartuples:\n",
    "                if type_indel == 1:\n",
    "                    lenght_insertion.append(lenght)\n",
    "                if type_indel == 2:\n",
    "                    lenght_deletion.append(lenght)\n",
    "                if type_indel == 1 or type_indel == 2:\n",
    "                    lenght_indel.append(lenght)\n",
    "\n",
    "    tempory_counters = [Counter(array) for array in [lenght_insertion, lenght_deletion]]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(15, 10))\n",
    "    for counter, label in zip(tempory_counters, ['insertion', 'deletion']):\n",
    "        x = list(counter.keys())\n",
    "        y = list(counter.values())\n",
    "        plt.scatter(x, y, label=label, linewidth=2)\n",
    "\n",
    "    plt.legend(loc=1, prop={'size': 16})\n",
    "    plt.xlabel('Длина индела')\n",
    "    plt.ylabel('Количество инделов с такой длиной')\n",
    "    plt.grid()\n",
    "    plt.xlim(0, 100)\n",
    "\n",
    "def plot_kde_replacment(samfile):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"r\") \n",
    "    quality_indel = []\n",
    "    for read in samfile.fetch():\n",
    "        acnt, generator = -1, read.cigartuples\n",
    "        if generator:\n",
    "            for ind, cnt in generator:\n",
    "                if ind == 0:\n",
    "                    acnt += cnt\n",
    "                if ind == 1:\n",
    "                    quality_indel.append(ord(read.qual[acnt]) - 33)\n",
    "    \n",
    "    plt.figure(figsize=(15, 10))\n",
    "    plt.xlabel('Качество нуклеотида')\n",
    "    plt.ylabel('Частота')\n",
    "    sns.kdeplot(quality_indel)\n",
    "def template_lenght(samfile, confidience, scale=False):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"r\") \n",
    "    templates = []\n",
    "    for x in samfile.fetch():\n",
    "        if x.template_length > 0:\n",
    "            templates.append(x.template_length)\n",
    "            \n",
    "    samfile.close()\n",
    "    \n",
    "    counter_templates = Counter(templates)\n",
    "    x = np.array(list(range(len(counter_templates))))\n",
    "    y = np.array([counter_templates[i] for i in range(len(counter_templates))])\n",
    "    if scale:\n",
    "        x, y = x[:1000], y[:1000]\n",
    "        \n",
    "    #draw temlates lenght\n",
    "    plt.figure(figsize=(15, 10))\n",
    "    plt.xlabel('Расстояние вставки')\n",
    "    plt.ylabel('Количество с таким расстоянием вставки')\n",
    "    plt.grid()\n",
    "    \n",
    "    # Calculate cond interval\n",
    "    cnt, min_index_l, min_index_r = 0, 0, 0\n",
    "    left, right = sum(y) * (1 - confidience) / 2, sum(y) * (1 + confidience) / 2\n",
    "    for index, elem in enumerate(array[1]):\n",
    "        cnt += elem\n",
    "        if cnt > left and min_index_l == 0:\n",
    "            min_index_l = index\n",
    "        if cnt > right and min_index_r == 0:\n",
    "            min_index_r = index\n",
    "            break\n",
    "    \n",
    "    #draw other\n",
    "    plt.axvline(x[min_index_l], linestyle='--', color='r', label='left confidence interval')\n",
    "    plt.axvline(x[min_index_r], linestyle='--', color='r', label='right confidence interval')\n",
    "    plt.legend()\n",
    "    sns.scatterplot(x, y, color='k')\n",
    "    return x, y\n",
    "\n",
    "def matrix_substitutions(samfile, refrence):\n",
    "    samfile = pysam.AlignmentFile(samfile, \"rb\")\n",
    "    replacment = {'A' : [], 'G' : [], 'C' : [], 'T' : [], '_' : []}\n",
    "    refseq = SeqIO.read(refrence, 'fasta')\n",
    "    \n",
    "    for read in samfile.fetch():\n",
    "        generator, acc = [], 0\n",
    "        if read.cigartuples and read.seq:\n",
    "            for ind, count in read.cigartuples:\n",
    "                generator.append([ind, [acc, acc + count]])\n",
    "                acc += count\n",
    "            l_read, l_reference, counter_ind, cnt = 0, read.reference_start, 0, 0\n",
    "            iterations = 0\n",
    "            while l_read != len(read.seq):\n",
    "                if generator[counter_ind][0] == 0:\n",
    "                    replacment[refseq[l_reference]].append(read.seq[l_read])\n",
    "                    l_read += 1\n",
    "                    l_reference += 1\n",
    "                if generator[counter_ind][0] == 1:\n",
    "                    replacment['_'].append(read.seq[l_read])\n",
    "                    l_reference += 1\n",
    "                    l_read += 1\n",
    "                if generator[counter_ind][0] == 2:\n",
    "                    replacment[refseq[l_reference]].append('_')\n",
    "                    l_read += 1\n",
    "                if generator[counter_ind][0] == 4:\n",
    "                    l_read += 1\n",
    "                    l_reference += 1\n",
    "                if l_read == generator[counter_ind][1][1]:\n",
    "                    counter_ind += 1\n",
    "                if iterations > 150:\n",
    "                    break\n",
    "    samfile.close()\n",
    "    \n",
    "    return replacment\n",
    "\n",
    "def matrix_substitutions_2(samfile, refrence):\n",
    "    samfile = pysam.AlignmentFile(samfile, threads=4)\n",
    "    refseq = SeqIO.read(refrence, 'fasta')\n",
    "    matrix = {nucl_1 : {nucl_2 : 0 for nucl_2 in ['A', 'G', 'C', 'T', '']} for nucl_1 in ['A', 'G', 'C', 'T']}\n",
    "    for columns in samfile.pileup():\n",
    "        ref_nucl = refseq[columns.reference_pos]\n",
    "        for nucl in ['A', 'C', 'T', 'G', '']:\n",
    "            matrix[ref_nucl][nucl] += Counter(columns.get_query_sequences())[nucl]\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "BAM_FILE_RAW = 'aln_sorted.bam'\n",
    "BAM_FILE_TRIMMOMATIC = 'aln_sorted_tr.bam'\n",
    "REFERENCE_FILE = 'MG1655-K12.first400K.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "240000\n",
      "250000\n",
      "260000\n",
      "270000\n",
      "280000\n",
      "290000\n",
      "300000\n",
      "310000\n",
      "320000\n",
      "330000\n",
      "340000\n",
      "350000\n",
      "360000\n",
      "370000\n",
      "380000\n",
      "390000\n",
      "400000\n"
     ]
    }
   ],
   "source": [
    "bamfile_raw = pysam.AlignmentFile(BAM_FILE_RAW, threads=4)\n",
    "bamfile_trimmomatic = pysam.AlignmentFile(BAM_FILE_TRIMMOMATIC, threads=4)\n",
    "refseq = SeqIO.read(REFERENCE_FILE, 'fasta')\n",
    "count = 0\n",
    "matrix_replacment = {'Undetected' : 0, 'Falsely corrected' : 0, 'Detected corrected' : 0, \n",
    "                     'Correctly unmodified' : 0, 'Detected and removed' : 0, 'Incorrectly removed' : 0}\n",
    "for columns_1, columns_2 in zip(bamfile_raw.pileup(), bamfile_trimmomatic.pileup()):\n",
    "    ref_nucl = refseq[columns_1.reference_pos]\n",
    "    names_reads = {}\n",
    "    \n",
    "    for read_corr in columns_2.pileups:\n",
    "        if read_corr.query_position is not None:\n",
    "            names_reads[read_corr.alignment.query_name] = read_corr.alignment.query_sequence[read_corr.query_position]\n",
    "        else:\n",
    "            names_reads[read_corr.alignment.query_name] = ''\n",
    "    \n",
    "    for read_raw in columns_1.pileups:\n",
    "        if read_raw.alignment.query_name in names_reads:\n",
    "            if read_raw.query_position:\n",
    "                raw_nucl  = read_raw.alignment.query_sequence[read_raw.query_position]\n",
    "                corr_nucl = names_reads[read_raw.alignment.query_name]\n",
    "                if (raw_nucl == corr_nucl) and corr_nucl != ref_nucl:\n",
    "                    matrix_replacment['Undetected'] += 1\n",
    "\n",
    "                if (raw_nucl == ref_nucl) and corr_nucl != ref_nucl:\n",
    "                    matrix_replacment['Falsely corrected'] += 1\n",
    "\n",
    "                if (raw_nucl != ref_nucl) and corr_nucl == ref_nucl:\n",
    "                    matrix_replacment['Detected corrected'] += 1\n",
    "\n",
    "                if raw_nucl == ref_nucl and corr_nucl == ref_nucl:\n",
    "                    matrix_replacment['Correctly unmodified'] += 1\n",
    "        else:\n",
    "            if read_raw.query_position:\n",
    "                raw_nucl  = read_raw.alignment.query_sequence[read_raw.query_position]\n",
    "                if raw_nucl == ref_nucl:\n",
    "                    matrix_replacment['Incorrectly removed'] += 1\n",
    "                else:\n",
    "                    matrix_replacment['Detected and removed'] += 1\n",
    "    count += 1\n",
    "    if count % 10000 == 0:\n",
    "        print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Undetected': 421002,\n",
       " 'Falsely corrected': 155,\n",
       " 'Detected corrected': 219,\n",
       " 'Correctly unmodified': 239846944,\n",
       " 'Detected and removed': 22007,\n",
       " 'Incorrectly removed': 6582739}"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_replacment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
